src/AbstractApiClient.php

```diff
@@ -0,0 +1,266 @@
+<?php declare(strict_types=1);
+/*
+ * This file is part of the Bushido\ApiClient package.
+ *
+ * (c) Wojciech Nowicki <wnowicki@me.com>
+ *
+ * For the full copyright and license information, please view the LICENSE.md
+ * file that was distributed with this source code.
+ */
+
+namespace Bushido\ApiClient;
+
+use Bushido\ApiClient\Exceptions\BadResponseException;
+use Bushido\ApiClient\Exceptions\ErrorResponseException;
+use Bushido\ApiClient\Exceptions\TransportException;
+use Bushido\ApiClient\Exceptions\WrongResponseException;
+use Bushido\Foundation\Helpers\PsrLoggerTrait;
+use GuzzleHttp\Client;
+use GuzzleHttp\ClientInterface;
+use GuzzleHttp\Exception;
+use GuzzleHttp\Exception\GuzzleException;
+use GuzzleHttp\Psr7\Request;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Log\LoggerInterface;
+
+abstract class AbstractApiClient
+{
+    use PsrLoggerTrait;
+
+    private $client;
+    private $logger;
+    private $headers;
+
+    public function __construct(array $config = [], LoggerInterface $logger = null, array $headers = [])
+    {
+        $this->setClient(new Client($config));
```

Comment:
Normally I would freak out on this, but there are functional tests, so it's fine. (See more on that later)

Reply:
Please bear in mind that we are not actually recreating here Guzzle functionality, we are providing another layer of abstraction for communication content.

Reply:
It's just the tight coupling that makes me alerted, but so long it is tested, it's not an issue.